colossalai/utils/common.py

```diff
@@ -161,8 +162,95 @@ def _get_tensor_norm(norm: Union[float, torch.Tensor], move_to_cuda) -> torch.Te
 
 # ======== Gradient Clipping =========
 
+def _norm_tensor_parallel_fp32(parameters, norm_type=2):
+    '''
+    Calculate norm.
+    Won't check if parameters are TP only
+    expect parameter grads on cuda
+    '''
+    if isinstance(parameters, ColoTensor):
+        parameters = [parameters]
+    params: List[ColoParameter] = []
+    for param in parameters:
+        if param.grad is not None:
+            assert param.grad.dtype == torch.float, \
+                f'expected gradient to be dtype torch.float, but got {param.grad.type()}'
+            assert param.grad.device.type == 'cuda', \
+                f'expected gradient to be on cuda, but got {param.grad.device.type}'
+            params.append(param)
+    norm_type = float(norm_type)
+    total_norm = 0.0
+    if norm_type == inf:
+        for p in params:
+            local_norm = p.grad.data.abs().max()
+            local_norm_cuda = torch.cuda.FloatTensor([float(local_norm)])
+            if p.is_sharded():
+                dist.all_reduce(local_norm_cuda,
+                                op=dist.ReduceOp.MAX,
+                                group=p.get_process_group().tp_process_group(),
+                                async_op=False)
+            total_norm = max(total_norm, local_norm_cuda[0].item())
+
+    else:
+        for p in params:
+            if norm_type == 2.0:
+                local_norm = _calc_l2_norm([p.grad.data])**norm_type
+            else:
+                local_norm = _calc_lp([p.grad.data], norm_type)
+            local_norm_cuda = torch.cuda.FloatTensor([float(local_norm)])
+            if p.is_sharded():
+                dist.all_reduce(local_norm_cuda,
```

Comment:
This leads to many unexpected communications. You may use a bucket.

Reply:
done